src/style-spec/function/type_check.js

```diff
@@ -0,0 +1,300 @@
+// tree with all generics resolved
+function typeCheckExpression(expected: Type, e: TypedExpression) /*: TypedExpression | {| errors: Array<TypeError> |} */ {
+    if (e.literal) {
+        const error = match(expected, e.type);
+        if (error) return { errors: [{ key: e.key, error }] };
+        return e;
+    } else {
+        // e is a lambda expression, so check its result type against the
+        // expected type and recursively typecheck its arguments
+
+        const typenames: { [string]: Type } = {};
+
+        if (expected.kind !== 'lambda') {
+            // if the expected type is not a lambda, then check if it matches
+            // the expression's output type, and then proceed with type checking
+            // the arguments using e.type, which comes from the expression
+            // definition.
+            const error = match(expected, e.type.result, {}, typenames);
+            if (error) return { errors: [{ key: e.key, error }] };
+            expected = e.type;
+        } else {
+            const error = match(expected.result, e.type.result, typenames);
+            if (error) return { errors: [{ key: e.key, error }] };
+        }
+
+        // Iterate through arguments to:
+        //  - match expected argument type vs argument type, checking result type only -- don't recursively typecheck subexpressions at this stage
+        //  - collect typename mappings when ^ succeeds or type errors when it fails
+        //  - "unroll" NArgs by greedily matching them against argument values,
+        //    building up an NArgs-free `expandedArgTypes` list in the process.
+        const argValues = e.arguments;
+        const expandedArgTypes = [];
+        const errors = [];
+        for (let vi = 0, ti = 0; vi < argValues.length && ti < expected.args.length; ti++) {
+            const t = expected.args[ti];
+            if (t.kind === 'nargs') {
+                let sequenceCount = 0;
+                let j = 0;
+                let argTypeNames = {};
+                while (sequenceCount < t.N && vi < argValues.length) {
+                    const error = match(
+                        resolveTypenamesIfPossible(t.types[j], typenames),
+                        argValues[vi].type,
+                        argTypeNames
+                    );
+                    if (error) console.log(error)
+                    if (error) break;
+
+                    vi++;
+                    j = (j + 1) % t.types.length;
+                    // If we've consumed a full sequence, append a copy of this
+                    // list of types to the expanded argument list
+                    if (j === 0) {
+                        sequenceCount++;
+                        expandedArgTypes.push.apply(expandedArgTypes, t.types);
+                        util.extend(typenames, argTypeNames);
+                        argTypeNames = {};
+                    }
+                }
+                vi -= j;
```

Comment:
This loop/logic could be simplified if we make the (pretty reasonable) restriction that an expression can only have one `NArgs` pack in its parameter list.  With that restriction, we could simply calculate the number of times we need to repeat the NArgs parameter pack to form an unrolled parameter list that matches the number of arguments actually provided.